### ISSUE TITLE
Overloads don't work on decorated functions

### ISSUE THREAD

**Issue by ncollins-vs** [astral-sh/ty#2278]
### Summary

```python
from typing import overload, ParamSpec, TypeVar, Callable
import functools

PS = ParamSpec("PS")
T = TypeVar("T")
def decorator(f: Callable[PS, T]) -> Callable[PS, T]:
    @functools.wraps(f)
    def wrapper(*args: PS.args, **kwargs: PS.kwargs) -> T:
        return f(*args, **kwargs)
    return wrapper


@overload
def test(x: int) -> int: ...
@overload
def test(*, y: str) -> str: ...
@decorator
def test(x: int | None = None, *, y: str | None = None) -> int | str:
    raise NotImplementedError

x = 1
reveal_type(x)
reveal_type(test(x))
```

If I comment out `@decorator`, ty correctly determines that `test(x)` is of type `int`. But with the decorator it incorrectly determines the type as `int | str` as if the overloads didn't exist.

[Playground link](https://play.ty.dev/24787475-a3a3-4ba9-9450-4bcdebc05f92)

### Version

ty 0.0.8 (aa7559db8 2025-12-29)

**Comment by carljm** [comment]
Thanks for the report! Definitely a bug.

**Comment by Dheebz** [comment]
Looking into this one I think i might have it figured out.

I'll just do some additional documentation and improve the unit tests a bit and happy to raise a PR.

- Fix: Preserve overloads when applying a decorator by re-invoking the decorator on each
  overload signature and rebuilding an overloaded callable if the decorator returns a
  callable; fall back to previous behavior when the return type can’t be expressed as a
  callable. This keeps overload return-type precision (e.g., test(1) -> int).
- Added test: call/function.md mdtest case for a ParamSpec-preserving decorator applied to
  an overloaded function; reveal_type(test(1)) -> int, reveal_type(test(y="hi")) -> str.

- ruff/crates/ty_python_semantic/src/types/infer/builder.rs
- ruff/crates/ty_python_semantic/resources/mdtest/call/function.md

**Comment by aidandj** [comment]
Found a possibly related case, can open a new issue if it is different. When using overloads, I'm getting invalid argument errors when also using a decorator.

https://play.ty.dev/e7a7f281-da81-466d-ac99-e4f85ad38d4e

```
from contextlib import contextmanager
from typing import overload, Dict, Type, Iterator, TypeVar

class Base: ...

class Child(Base): ...

T = TypeVar("T", bound=Base)

@overload
@contextmanager
def begin(locks: Dict[Type[T], int]) -> Iterator[int]: ...

@overload
@contextmanager
def begin(locks: Dict[str, int]) -> Iterator[str]: ...

@contextmanager
def begin(locks: Dict[Type[T], int] | Dict[str, int]) -> Iterator[str] | Iterator[int]:
    yield 1

begin({Child: 1})

```

**Comment by carljm** [comment]
It looks likely related to me, since removing / commenting the `@contextmanager` decorator(s) in that example makes the error go away and bidirectional inference then handles the dict literal argument fine.

**Comment by jakob-keller** [comment]
I figure this is the root cause for a number of false positive detections when I update my project's polars dependency to 1.38.0.

**Comment by Jeremiah-England** [comment]
> I figure this is the root cause for a number of false positive detections when I update my project's polars dependency to 1.38.0.

Yes, this is breaking `LazyFrame.collect()`.

In v0.0.14, this wasn't much of an issue for me, because `ty` was not able to resolve `LazyFrame.collect()` at all and returned `Unknown`. In v0.0.15 it returns an incorrect union type, causing a bunch of downstream type issues since `InProcessQuery` cannot be used as a `DataFrame`.

<details>
<summary>
example.py
</summary>

```sh
> uv add ty==0.0.14 --dev; uv run ty check example.py
example.py:36:13: info[revealed-type] Revealed type: `Unknown`
Found 1 diagnostic
> uv add ty==0.0.15 --dev; uv run ty check example.py
example.py:36:13: info[revealed-type] Revealed type: `InProcessQuery | DataFrame`
Found 1 diagnostic
```

```python
"""
A somewhat minimized version of the polars type issue using original Polar names.
"""

from collections.abc import Callable
from typing import Literal, ParamSpec, TypeVar, overload, reveal_type

P = ParamSpec("P")
T = TypeVar("T")


def deprecate_streaming_parameter() -> Callable[[Callable[P, T]], Callable[P, T]]:
    return ...  # ty: ignore[invalid-return-type]


class InProcessQuery: ...


class DataFrame: ...


class LazyFrame:
    @overload
    def collect(self, *, background: Literal[True]) -> InProcessQuery: ...

    @overload
    def collect(self, *, background: Literal[False] = False) -> DataFrame: ...

    @deprecate_streaming_parameter()
    def collect(self, *, background: bool = False) -> DataFrame | InProcessQuery:
        return ...  # ty: ignore[invalid-return-type]


# Return type here is InProcessQuery | DataFrame instead of just DataFrame
# However, if I comment out the decorator, I get just DataFrame.
reveal_type(LazyFrame().collect())
```

</details>

**Comment by carljm** [comment]
I think @dhruvmanila is working on this!

**Comment by dhruvmanila** [comment]
The issue here is not related to overloads but it's about `ParamSpec` and somewhat related to https://github.com/astral-sh/ty/issues/2799. The union is coming from the fact that the old solver would union multiple types for `T` which, in this case, is coming from the return types of the overloads. This gets passed on to the return type of `decorator` which is the `Callable[PS, T]`. A simple fix for the time being could be to recognize this pattern of using `Callable[P, T]`, retain the return types from the ParamSpec specialized type, and pass it on to the callable return type which should also be `Callable[P, T]`. I've a PR up (https://github.com/astral-sh/ruff/pull/23371) that does something similar but I don't think it's the correct solution (given the ecosystem changes). I'm planning to look into a different solution instead.

**Comment by gaborbernat** [comment]
## Real-world use case: polars LazyFrame.collect()

This issue directly affects the polars library. The `LazyFrame.collect()` method has multiple overloads with Literal type discrimination, but they don't work due to decorators:

```python
@deprecate_streaming_parameter()
@forward_old_opt_flags()
def collect(
    self,
    *,
    background: bool = False,
) -> DataFrame | InProcessQuery:
    ...
```

The overloads are:
- `background: Literal[True]` -> `InProcessQuery`
- `background: Literal[False] = False` -> `DataFrame`

But with the decorators, calling `lf.collect(background=False)` returns `DataFrame | InProcessQuery` instead of narrowing to `DataFrame`.

This affects any codebase migrating from pyright to ty and using polars).

Related issue: https://github.com/astral-sh/ty/issues/3141

**Comment by dhruvmanila** [comment]
Thanks for providing additional examples, this will be my top priority next week given that #1535 is complete (there are some follow-up issues that I'm currently working on).

**Comment by mdecuir** [comment]
I just ran into the exact same behavior with polars LazyFrame.collect() today using ty 0.0.32. Glad to see it is already engaged.

**Comment by Crypto-Spartan** [comment]
does anyone have a decent workaround for `LazyFrame.collect()`? I've just been using casts everywhere like this, which is less than ideal
```python
from typing import cast
import polars as pl

df = cast(pl.DataFrame, lf.collect())
```

**Comment by mflova** [comment]
> does anyone have a decent workaround for `LazyFrame.collect()`? I've just been using casts everywhere like this, which is less than ideal
> ```python
> from typing import cast
> import polars as pl
> 
> df = cast(pl.DataFrame, lf.collect())
> ```

This issue is labeled as one of the only ones missing before the pre stable milestone (there are like 25 left). If I were you, I would wait. Otherwise the other option might be ignoring lal type hints from that package (not sure if possible).

**Comment by abgros** [comment]
> does anyone have a decent workaround for `LazyFrame.collect()`? I've just been using casts everywhere like this, which is less than ideal

@Crypto-Spartan try this:
```py
df = pl.DataFrame(lf.collect())
```

**Comment by Crypto-Spartan** [comment]
I honestly don't think that is much better

**Comment by mflova** [comment]
> > does anyone have a decent workaround for `LazyFrame.collect()`? I've just been using casts everywhere like this, which is less than ideal
> 
> @Crypto-Spartan try this:
> ```py
> df = pl.DataFrame(lf.collect())
> ```

When doing that you are creating a whole copy of the dataframe just to deal with a type hint issue. Depending on the use case this can introduce quite a bit of overhead

**Comment by mdecuir** [comment]
> > does anyone have a decent workaround for `LazyFrame.collect()`? I've just been using casts everywhere like this, which is less than ideal
> > from typing import cast
> > import polars as pl
> > 
> > df = cast(pl.DataFrame, lf.collect())
> 
> This issue is labeled as one of the only ones missing before the pre stable milestone (there are like 25 left). If I were you, I would wait. Otherwise the other option might be ignoring lal type hints from that package (not sure if possible).

The approach my team is taking is just adding an appropriate `# ty: ignore` comment and explicitly type hinting for the result along the lines of:
```
df: pl.DataFrame = lf.collect(background=False)  # ty:ignore[invalid-assignment]
```
with the idea that the explicit type hinting of the variable will always be valid, and when this issue with ty is resolved, the ignore comments will show up as warnings for being unnecessary and we can remove them at that time.

**Comment by abgros** [comment]
> When doing that you are creating a whole copy of the dataframe just to deal with a type hint issue. Depending on the use case this can introduce quite a bit of overhead

@mflova you are wrong, and this can easily be shown using a benchmark. There is no measureable difference!

```py
df = pl.DataFrame({"foo": [time() for i in range(10000000)]})

elapsed1 = 0
elapsed2 = 0

for i in range(1000):
    start = perf_counter_ns()
    df2 = df.lazy().select(foo=pl.col("foo") * 2).collect()
    elapsed1 += perf_counter_ns() - start

    start = perf_counter_ns()
    df2 = pl.DataFrame(df.lazy().select(foo=pl.col("foo") * 2).collect())
    elapsed2 += perf_counter_ns() - start

elapsed1 /= 1e9
elapsed2 /= 1e9

print(f"Elapsed (1): {elapsed1:.2f}s")
print(f"Elapsed (2): {elapsed2:.2f}s")
print(f"(2) is {(elapsed2 - elapsed1) / elapsed1 * 100:.2f}% slower")
```

Results:
```
Elapsed (1): 16.79s
Elapsed (2): 16.84s
(2) is 0.31% slower
```

**Comment by dhruvmanila** [comment]
Hi, we've merged https://github.com/astral-sh/ruff/pull/25030 which is a temporary workaround for this and related issues, I'm going to keep this open to make sure the workaround is removed once #3555 is completed.

**Comment by albireox** [comment]
With 0.0.46 on VSCode I'm still seeing the same issue when calling `.collect()` on a Polars LazyFrame.

**Comment by dhruvmanila** [comment]
@albireox Thanks for getting back, I can take a quick look at that.

**Comment by dhruvmanila** [comment]
Yeah, the workaround doesn't fix the Polars issue because that method uses `Protocol` with a `__call__` method, I think we can expand the workaround to use `try_upcast_to_callable`. I'll get that going in the background to see if it's helpful.

**Comment by albireox** [comment]
Thanks for the quick fix. Using VSCode with the environment version 0.0.47 all instances that I was seeing before seems to work fine now.

**Comment by carljm** [comment]
This was fixed by [#27147](https://github.com/astral-sh/ruff/pull/27147)

**Comment by carljm** [comment]
@dhruvmanila I see https://github.com/astral-sh/ruff/pull/27235 is marked as closing this. That looks like a good cleanup PR, but it looks to me like all examples given in this issue are already fixed on main, so I'm going ahead and closing this.
